rfd/0163-vnet.md

```diff
@@ -0,0 +1,970 @@
+It is built on top of Teleport's security principals like short-lived
+certificates, device trust, passwordless authentication, and MFA.
+
+It will not be possible for external traffic to access the host over VNet.
+The only IP packets that VNet will send to the host OS will be replies to
+outgoing TCP connections that have been made to Teleport apps, and DNS answers.
+
+#### DNS rebinding
+
+Because VNet will expose Teleport protected apps on the local machine without
+any per-connection client authentication, we need to be careful that they cannot
+be reached by untrusted code.
+
+One source of untrusted code that often runs on a user's machine is Javascript
+executed in the web browser.
+Web browsers apply CORS protections to ensure that Javascript cannot access
+arbitrary web addresses outside the domain that the JS code originated from.
+However, CORS does not protect plain HTTP websites from DNS rebinding attacks,
+where the attacker's website dynamically changes its IP address to the address
+of a website it wishes to make illegitimate HTTP requests to.
+
+Here's an example scenario:
+
+- The TCP app `api.teleport.example.com` is listening for plain HTTP connections.
+- VNet assigns the address `100.64.0.2` to the app.
+- The user clicks a link to `hacker.example.com` which first resolves to a
+  public IP address with a very short TTL.
+- The web browser loads and executes the Javascript from `hacker.example.com`.
+- `hacker.example.com` "rebinds" its DNS address to `100.64.0.2`.
+- The Javascript for `hacker.example.com` makes a POST request to
+  `http://hacker.example.com/path`.
+- The browser will resolve the address of `hacker.example.com` to `100.64.0.2`
+  and the POST request will actually be sent to the `api.teleport.example.com`
+  app, authenticated as the user!
+
+This attack vector will only be relevant to apps listening for plain,
+unauthenticated HTTP connections that have been enrolled as TCP apps in the
+Teleport cluster (with a `spec.uri` prefix of `tcp://`).
+Hopefully this will be a very limited scenario.
+
+Normally, apps listening for HTTP that are meant to be opened in a web browser
+should be enrolled as web apps with a `spec.uri` prefix of `http://`, these
+won't be affected because VNet will not expose them and they will only be
+accessible over HTTPS.
+However, it could be desirable to access REST APIs over VNet, in this case they
+must be enrolled as a TCP app if they are to be accessed by non-browser clients.
+
+In case a Teleport admin does choose to enroll a plain HTTP app in their cluster as
+a TCP app, there are a few possible mitigations for this DNS rebinding attack
+scenario.
+
+1. Apply `Host` header allowlists for HTTP apps.
+2. Make VNet the global DNS nameserver on the host and reject any forwarded
+   queries that resolve to a VNet address or CNAME for a VNet name.
+3. Use a private IP range for VNet IPv4 addresses.
+4. Drop IPv4 and only support IPv6.
+
```

Comment:
Which of these mitigations has actually been chosen? This paragraph feels unconclusive.

I personally like the `Host` header allowlist, as it nicely uses HTTP protocol feature. The obvious downside is that the VNet feature has no way to actually enforce it (unless we would reconsider parsing HTTP requests by the VNet service). In my opinion it's fine though: Unauthenticated and unencrypted HTTP services should be discouraged anyway.

Reply:
Added the following:

> In conclusion, the chosen mitigations for DNS rebinding are:
>
> 1. Discourage users from registering plain HTTP apps as TCP apps that will be
>    made accessible over VNet. HTTPs should be used instead where possible.
> 2. Encourage users to implement a `Host` header allowlist if they most expose a
>    plain HTTP app through VNet.
> 3. Wait for browser Private Network Access mitigations to recognize the CGNAT
>    range as a private IP space.

Reply:
Looks good, thx @nklaassen